src/app/identity.rs

```diff
@@ -0,0 +1,220 @@
+}
+
+// === impl Config ===
+
+impl Config {
+    fn refresh(&self, expiry: SystemTime) -> Delay {
+        let now = clock::now();
+
+        let refresh = match expiry
+            .duration_since(SystemTime::now())
+            .ok()
+            .map(|d| d * 7 / 10)
+        {
+            None => self.min_refresh,
+            Some(lifetime) if lifetime < self.min_refresh => self.min_refresh,
+            Some(lifetime) if self.max_refresh < lifetime => self.max_refresh,
+            Some(lifetime) => lifetime,
+        };
+
+        Delay::new(now + refresh)
+    }
+}
+
+// === impl Local ===
+
+impl Local {
+    pub fn new(config: &Config) -> (Self, CrtKeyStore) {
+        let (w, s) = Watch::new(None);
+        let l = Local {
+            name: config.local_name.clone(),
+            trust_anchors: config.trust_anchors.clone(),
+            crt_key: w,
+        };
+        (l, s)
+    }
+}
+
+impl tls::client::HasConfig for Local {
+    fn tls_client_config(&self) -> Arc<tls::client::Config> {
+        use transport::tls::client::HasConfig;
+
+        if let Some(ref c) = *self.crt_key.borrow() {
+            return c.tls_client_config();
+        }
+
+        self.trust_anchors.tls_client_config()
+    }
+}
+
+impl tls::listen::HasConfig for Local {
+    fn tls_server_name(&self) -> Name {
+        self.name.clone()
+    }
+
+    fn tls_server_config(&self) -> Arc<tls::listen::Config> {
+        use transport::tls::listen::HasConfig;
+
+        if let Some(ref c) = *self.crt_key.borrow() {
+            return c.tls_server_config();
+        }
+
+        tls::listen::empty_config()
+    }
+}
+
+// === impl Daemon ===
+
+impl<T> Daemon<T>
+where
+    T: GrpcService<BoxBody> + Clone,
+{
+    pub fn new(config: Config, crt_key: CrtKeyStore, client: T) -> Self {
+        Self {
+            config,
+            crt_key,
+            inner: Inner::ShouldRefresh,
+            expiry: UNIX_EPOCH,
+            client: api::client::Identity::new(client),
+        }
+    }
+}
+
+impl<T> Future for Daemon<T>
+where
+    T: GrpcService<BoxBody> + Clone,
+{
+    type Item = ();
+    type Error = Never;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            self.inner = match self.inner {
+                Inner::Waiting(ref mut d) => {
+                    if let Ok(Async::NotReady) = d.poll() {
+                        return Ok(Async::NotReady);
+                    }
+                    Inner::ShouldRefresh
+                }
+                Inner::ShouldRefresh => match self.config.token.load() {
+                    Ok(token) => {
+                        let req = grpc::Request::new(api::CertifyRequest {
+                            token,
+                            identity: self.config.local_name.as_ref().to_owned(),
+                            certificate_signing_request: self.config.csr.to_vec(),
+                        });
+                        let f = self.client.certify(req);
```

Comment:
Recently, the poll-ready-before-call contract was made strong in tower. While there's likely a `buffer` service protecting against most cases, many services will now *panic* instead return an error future if called and they weren't ready...

Reply:
This service is intended to be buffered, though it's not documented in the type system, so I agree we should poll-ready